### PR TITLE
irmin: add Repo.iter

### PR DESCRIPTION
**Pull request by samoht**
It allows to iterate efficiently over the object graph.

Extracted from #1124

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -52,6 +52,8 @@
   - Added `iter_commits` and `iter_nodes` functions to traverse the commits and
     nodes graphs (#1077, @icristescu)
 
+  - Adeed `Repo.iter` to traverse the object graphs (#1128, @samoht)
```
Comment (CraigFe) [inline CHANGES.md:55]:
```suggestion
  - Added `Repo.iter` to traverse object graphs (#1128, @samoht)
```

File: src/irmin/object_graph.ml [hunk 2]
```diff
@@ -138,6 +132,13 @@ struct
     let mark key level = Table.add marks key level in
     let has_mark key = Table.mem marks key in
     let todo = Stack.create () in
+    (* if a branch is in [min], add the commit it is pointing to it too. *)
```
Comment (CraigFe) [inline src/irmin/object_graph.ml:135]:
```suggestion
    (* if a branch is in [min], add the commit it is pointing to too. *)
```

File: src/irmin/s.ml [hunk 3]
```diff
@@ -360,10 +360,10 @@ module type NODE_GRAPH = sig
     min:node list ->
     max:node list ->
     ?node:(node -> unit Lwt.t) ->
-    ?contents:(contents * metadata -> unit Lwt.t) ->
+    ?contents:(contents -> unit Lwt.t) ->
     ?edge:(node -> node -> unit Lwt.t) ->
-    ?skip_nodes:(node -> bool Lwt.t) ->
-    ?skip_contents:(contents * metadata -> bool Lwt.t) ->
+    ?skip_node:(node -> bool Lwt.t) ->
```
Comment (CraigFe) [inline src/irmin/s.ml:365]: GitHub won't let me suggest it, but need to `s/skip_nodes/skip_node/` in comment below.

File: src/irmin/store_intf.ml [hunk 4]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+        [skip_branch], [skip_commit], [skip_node] and [skip_contents] allow the
+        traversal to be stopped when the corresponding objects are traverse. By
+        default no objects is skipped.
+
+        The [branch], [commit], [node] and [contents] functions are called
+        whenever the corresponding objects are traversed. By default these
+        functions do nothing. These functions are not called on skipped objects.
+
+        [pred_branch], [pred_commit], [pred_node] and [pred_contents]
+        implicitely define the graph underlying the traversal. By default they
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:163]:
```suggestion
        implicitly define the graph underlying the traversal. By default they
```

File: src/irmin/store_intf.ml [hunk 5]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+        exactly match the underlying Merkle graph of the repository [t]. These
+        functions can be used to traverse a sligthly modified version of that
+        graph, for instance by modifying [pred_contents] to implicitely link
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:166]:
```suggestion
        graph, for instance by modifying [pred_contents] to implicitly link
```

File: src/irmin/store_intf.ml [hunk 6]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+        default no objects is skipped.
+
+        The [branch], [commit], [node] and [contents] functions are called
+        whenever the corresponding objects are traversed. By default these
+        functions do nothing. These functions are not called on skipped objects.
+
+        [pred_branch], [pred_commit], [pred_node] and [pred_contents]
+        implicitely define the graph underlying the traversal. By default they
+        exactly match the underlying Merkle graph of the repository [t]. These
+        functions can be used to traverse a sligthly modified version of that
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:165]:
```suggestion
        functions can be used to traverse a slightly modified version of that
```

File: src/irmin/store_intf.ml [hunk 7]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+        graph, for instance by modifying [pred_contents] to implicitely link
+        structured contents with other objects in the graph.
+
+        The traversed objects are all included between in [min] (included) and
+        [max] (included), following the Merkle graph order. Moreover, The [min]
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:170]:
```suggestion
        [max] (included), following the Merkle graph order. Moreover, the [min]
```

File: src/irmin/store_intf.ml [hunk 8]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+        functions do nothing. These functions are not called on skipped objects.
+
+        [pred_branch], [pred_commit], [pred_node] and [pred_contents]
+        implicitely define the graph underlying the traversal. By default they
+        exactly match the underlying Merkle graph of the repository [t]. These
+        functions can be used to traverse a sligthly modified version of that
+        graph, for instance by modifying [pred_contents] to implicitely link
+        structured contents with other objects in the graph.
+
+        The traversed objects are all included between in [min] (included) and
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:169]:
```suggestion
        The traversed objects are all included between [min] (included) and
```

File: src/irmin/store_intf.ml [hunk 9]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+        boundary is extended as follows:
+
+        - contents and node objects in [min] stop the traversal; their
+          predecessors are not traversed.
+        - commit objects in [min] stop the traversal for their commit
+          predecessors, but their sub-node are still traversed. This allow users
+          to define an inclusive range of commit to iterate other.
+        - branch object in [min] implicitely add to [min] the commit they are
+          pointint to; This allow users to define the iteration between two
+          branches. *)
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:180]:
```suggestion
        - branch object in [min] implicitely add to [min] the commit they are
          pointint to; This allow users to define the iteration between two
          branches. *)
```
```suggestion
        - branch objects in [min] implicitly add to [min] the commit they are
          pointing to; this allow users to define the iteration between two
          branches. *)
```

File: src/irmin/store_intf.ml [hunk 10]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+
+        The traversed objects are all included between in [min] (included) and
+        [max] (included), following the Merkle graph order. Moreover, The [min]
+        boundary is extended as follows:
+
+        - contents and node objects in [min] stop the traversal; their
+          predecessors are not traversed.
+        - commit objects in [min] stop the traversal for their commit
+          predecessors, but their sub-node are still traversed. This allow users
+          to define an inclusive range of commit to iterate other.
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:177]:
```suggestion
          predecessors, but their sub-node are still traversed. This allows users
          to define an inclusive range of commit to iterate over.
```

File: src/irmin/store_intf.ml [hunk 11]
```diff
@@ -123,15 +123,71 @@ module type S = sig
...
+      ?pred_contents:(t -> hash -> elt list Lwt.t) ->
+      ?rev:bool ->
+      unit ->
+      unit Lwt.t
+    (** [iter t] iterates in topological order over the closure graph of [t]. If
+        [rev] is set (by default it is) the traversal is done in reverse order.
+
+        [skip_branch], [skip_commit], [skip_node] and [skip_contents] allow the
+        traversal to be stopped when the corresponding objects are traverse. By
+        default no objects is skipped.
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:156]:
```suggestion
        traversal to be stopped when the corresponding objects are traversed. By
        default no objects are skipped.
```

File: src/irmin/store_intf.ml [hunk 12]
```diff
@@ -123,15 +123,71 @@ module type S = sig
     (** [import t s] imports the contents of the slice [s] in [t]. Does not
         modify branches. *)
 
+    type elt =
+      [ `Commit of hash | `Node of hash | `Contents of hash | `Branch of branch ]
+    [@@deriving irmin]
+    (** The type for elements iterated over by [iter]. *)
```
Comment (CraigFe) [inline src/irmin/store_intf.ml:129]:
```suggestion
    (** The type for elements iterated over by {!iter}. *)
```

File: src/irmin/dot.ml [hunk 13]
```diff
@@ -202,12 +200,8 @@ module Make (S : Store.S) = struct
...
       bs
     >|= fun () ->
-    let map = function
-      | `Contents c -> `Contents (c, Node.Metadata.default)
-      | (`Commit _ | `Node _ | `Branch _) as k -> k
-    in
-    let vertex = Hashtbl.fold (fun k v acc -> (map k, v) :: acc) vertex [] in
-    let edges = List.map (fun (k, l, v) -> (map k, l, map v)) !edges in
+    let vertex = Hashtbl.fold (fun k v acc -> (k, v) :: acc) vertex [] in
+    let edges = List.map (fun (k, l, v) -> (k, l, v)) !edges in
```
Comment (CraigFe) [inline src/irmin/dot.ml:204]: Looks like we can remove this `map` altogether?

**Review by CraigFe — COMMENTED** [review]
LGTM. Would be good to have tests of the newly-introduced edge cases re. branches and cuts, but that can probably wait for another day.

**Review by icristescu — APPROVED** [review]
